lib/sql.ts

```diff
@@ -0,0 +1,164 @@
+/**
+ * Async worker thread wrapper around SQLite, written to improve concurrent performance.
+ * @author mia-pi-git
+ */
+import {parentPort, isMainThread} from 'worker_threads';
+import {ProcessWrapper} from './process-manager';
+import * as Sqlite from 'better-sqlite3';
+
+interface SQLOptions {
+	file: string;
+	/** file to import database functions from - this should be relative to this filename. */
+	extension?: string;
+}
+
+type DataType = unknown[] | Record<string, unknown>;
+
+export type DatabaseQuery = {
+	/** Prepare a statement - data is the statement. */
+	type: 'prepare', data: string,
+} | {
+	/** Get all lines from a statement. Data is the params, num is the statement number. */
+	type: 'all', data: DataType, num: number,
+} | {
+	/** Execute raw SQL in the database. */
+	type: "exec", data: string,
+} | {
+	/** Get one line from a prepared statement. */
+	type: 'get', data: DataType, num: number,
+} | {
+	/** Run a prepared statement. */
+	type: 'run', data: DataType, num: number,
+};
+
+if (!isMainThread) {
+	let statementNum = 0;
+	const statements: Map<number, Sqlite.Statement> = new Map();
+	const {file, extension} = process.env;
+	const database = new Sqlite(file!);
+	if (extension) {
+		// eslint-disable-next-line @typescript-eslint/no-var-requires
+		const functions: {[k: string]: (...params: any) => any} = require(`../${extension}`);
+		for (const key in functions) {
+			database.function(key, functions[key]);
+		}
+	}
+	database.pragma(`foreign_keys=on`);
+	parentPort!.on('message', (message: string) => {
```

Comment:
This does not appear to be actually using `ProcessManager`. Like I said, the value of using PM is that you can easily switch between subthreads, subprocesses, and synchronous execution. Hardcoding thread code defeats that goal.

You don't actually need to base on top of my PM code at all, you can just implement this as a PM subprocess, and then my PR will add thread support.

Reply:
Apologies! Misunderstood. Is this more what you meant, now?